modules/apps/29-fee/types/genesis_test.go

```diff
@@ -1,39 +1,197 @@
+		packetId     *channeltypes.PacketId
+		fee          types.Fee
+		refundAcc    string
+		sender       string
+		counterparty string
+		portID       = types.PortID
+		channelID    = ibctesting.FirstChannelID
+		seq          = uint64(1)
+	)
+
 	testCases := []struct {
 		name     string
-		genState *types.GenesisState
+		malleate func()
 		expPass  bool
 	}{
 		{
-			name:     "default",
-			genState: types.DefaultGenesisState(),
-			expPass:  true,
+			"valid genesis",
+			func() {},
+			true,
 		},
 		{
-			"valid genesis",
-			&types.GenesisState{
-				PortId: "portidone",
+			"invalid packetId: invalid channel",
+			func() {
+				packetId = types.NewPacketId(
+					"",
+					portID,
+					seq,
+				)
+
 			},
-			true,
+			false,
+		},
+		{
+			"invalid packetId: invalid port",
+			func() {
+				packetId = types.NewPacketId(
+					channelID,
+					"",
+					seq,
+				)
+			},
+			false,
+		},
+		{
+			"invalid packetId: invalid sequence",
+			func() {
+				packetId = types.NewPacketId(
+					channelID,
+					portID,
+					0,
+				)
+			},
+			false,
+		},
+		{
+			"invalid packetId: invalid fee",
+			func() {
+				fee = types.Fee{
+					sdk.Coins{},
+					sdk.Coins{},
+					sdk.Coins{},
+				}
+			},
+			false,
+		},
+		{
+			"invalid packetId: invalid refundAcc",
+			func() {
+				refundAcc = ""
+			},
+			false,
+		},
+		{
+			"invalid FeeEnabledChannel: invalid ChannelID",
+			func() {
+				channelID = ""
+			},
+			false,
+		},
+		{
+			"invalid FeeEnabledChannel: invalid PortID",
+			func() {
+				portID = ""
+			},
+			false,
 		},
 		{
-			"invalid client",
-			&types.GenesisState{
-				PortId: "(INVALIDPORT)",
+			"invalid RegisteredRelayers: invalid sender",
+			func() {
+				sender = ""
 			},
 			false,
 		},
+		{
+			"invalid RegisteredRelayers: invalid counterparty",
+			func() {
+				counterparty = ""
+			},
+			false,
+		},
+	}
+
+	for _, tc := range testCases {
+		portID = types.PortID
+		channelID = ibctesting.FirstChannelID
+		seq = uint64(1)
+
+		// build PacketId & Fee
+		packetId = types.NewPacketId(
+			portID,
+			channelID,
+			seq,
+		)
+		fee = types.Fee{
+			validCoins,
+			validCoins2,
+			validCoins3,
+		}
+
+		refundAcc = addr1
+
+		// relayer addresses
+		sender = addr1
+		counterparty = addr2
+
+		tc.malleate()
+
+		genState := types.GenesisState{
+			IdentifiedFees: []*types.IdentifiedPacketFee{
+				{
+					PacketId:      packetId,
+					Fee:           fee,
+					RefundAddress: refundAcc,
+					Relayers:      nil,
+				},
+			},
+			FeeEnabledChannels: []*types.FeeEnabledChannel{
+				{
+					PortId:    portID,
+					ChannelId: channelID,
+				},
+			},
+			RegisteredRelayers: []*types.RegisteredRelayerAddress{
+				{
+					Address:             sender,
+					CounterpartyAddress: counterparty,
+				},
+			},
+		}
+
+		err := genState.Validate()
+		if tc.expPass {
+			require.NoError(t, err, tc.name)
+		} else {
+			require.Error(t, err, tc.name)
+		}
+	}
+}
+
+func TestValidateDefaultGenesis(t *testing.T) {
```

Comment:
Separate test as it was clunky otherwise as I had to include the full genesis state in each test case.

Reply:
nit: why even create the `testCases` and for loop?

Just call `types.DefaultGenesisState.Validate()` and then assert the error is nil. 1-2 line test depending on readability

Reply:
Good point.